libtenzir/src/export_bridge.cpp

```diff
@@ -0,0 +1,355 @@
+            while (self->state().open_partitions
+                   < self->state().mode.parallel) {
+              ++self->state().open_partitions;
+              detail::weak_run_delayed(self, duration::zero(), [self] {
+                self->state().pop_partition();
+              });
+            }
+          }
+          TENZIR_ASSERT(self->state().unpersisted_events);
+          for (auto& slice : *self->state().unpersisted_events) {
+            if (slice.import_time() > max_import_time) {
+              self->state().add_events(std::move(slice),
+                                       event_source::unpersisted);
+            }
+          }
+          self->state().unpersisted_events.reset();
+          // In case we get zero partitions back from the catalog we need to
+          // already signal that we're done here.
+          if (self->state().buffer_rp.pending() and self->state().is_done()) {
+            self->state().buffer_rp.deliver(table_slice{});
+          }
+        },
+        [self](const caf::error& err) {
+          self->quit(
+            diagnostic::error(err)
+              .note("{} failed to retrieve candidates from catalog", *self)
+              .to_error());
+        });
+  }
+  return {
+    [self](table_slice& slice) -> caf::result<void> {
+      self->state().add_events(std::move(slice),
+                               self->current_sender()
+                                   == self->state().importer_address
+                                 ? event_source::live
+                                 : event_source::retro);
+      return {};
+    },
+    [self](atom::get) -> caf::result<table_slice> {
+      // Forbid concurrent requests.
+      TENZIR_ASSERT(not self->state().buffer_rp.pending());
+      if (self->state().is_done()) {
+        return table_slice{};
+      }
+      if (not self->state().buffer.empty()) {
+        auto slice = std::move(self->state().buffer.front());
+        self->state().buffer.pop();
+        TENZIR_ASSERT(slice.rows() > 0);
+        auto& metric = self->state().metrics[slice.schema()];
+        TENZIR_ASSERT(metric.queued >= slice.rows());
+        metric.emitted += slice.rows();
+        metric.queued -= slice.rows();
+        self->state().num_queued_total -= slice.rows();
+        self->state().try_pop_partition();
+        return slice;
+      }
+      self->state().buffer_rp = self->make_response_promise<table_slice>();
+      return self->state().buffer_rp;
+    },
+    [self](caf::exit_msg& msg) {
+      self->quit(std::move(msg.reason));
+    },
+  };
+}
+
+} // namespace
+
+auto spawn_bridge(caf::scheduled_actor& parent, expression expr,
+                  export_mode mode, filesystem_actor filesystem,
+                  metric_handler metrics_handler,
+                  std::unique_ptr<diagnostic_handler> diagnostics_handler)
+  -> export_bridge_actor {
+  // TODO: How to best expose this?
+  return parent.spawn<caf::linked>(make_bridge, std::move(expr), mode,
```

Comment:
The function should really be called `spawn_and_link_export_bridge` if it unconditionally links. This is otherwise not clear from the call site, and it's a really important detail.

Reply:
I would ideally like to expose a generic way to spawn the actor such that the user can configure how it is spawned, but without exposing `bridge_state` in the header. Maybe you know a way to do it (can't believe that it's not possible, that should be such a common thing, and who wants the state in their state?), but other I'll just go with the `spawn_and_link_export_bridge` suggestion.

Reply:
I don't know how to spawn without passing these as template arguments, unfortunately.